### PR TITLE
Updated our hash implementations to match getween go and C++.

### PR DESCRIPTION
**Pull request by AWoloszyn**
This sits on top of https://github.com/google/gapid/pull/1654.
Only the last change is relevant.

### CODE REVIEW

File: gapil/runtime/cc/map.inc [hunk 1]
```diff
@@ -25,6 +26,99 @@
...
+        v ^= rotate_right(v, 28);
+        v += shift_left(v, 31);
+        return v;
+    }
+};
+
+template<>
+struct gapil_hash<float> {
+    uint64_t operator() (const float& f) {
+        return gapil_hash<fixedValue>()(fixedValue{static_cast<uint64_t>(*reinterpret_cast<const uint32_t*>(&f))});
```
Comment (ben-clayton) [inline gapil/runtime/cc/map.inc]: I believe for this sort of thing you're supposed to use a union.

File: gapil/runtime/cc/map.inc [hunk 2]
```diff
@@ -25,6 +26,99 @@
...
+
+#define HASH_INTEGER(type) \
+template<> \
+struct gapil_hash<type> {\
+    uint64_t operator() (const type&  v) { \
+        return static_cast<uint64_t>(v); \
+    } \
+};
+
+HASH_INTEGER(bool)
```
Comment (ben-clayton) [inline gapil/runtime/cc/map.inc]: http://en.cppreference.com/w/cpp/types/is_integral ?

File: gapil/runtime/cc/map.inc [hunk 3]
```diff
@@ -25,6 +26,99 @@
...
+HASH_INTEGER(int16_t)
+HASH_INTEGER(uint32_t)
+HASH_INTEGER(int32_t)
+HASH_INTEGER(uint64_t)
+HASH_INTEGER(int64_t)
+
+#undef HASH_INTEGER
+
+template<typename T>
+struct gapil_hash<T*> {
```
Comment (ben-clayton) [inline gapil/runtime/cc/map.inc]: `gapil::hash`?
Maybe put into its own header?

File: gapil/runtime/cc/map.inc (hunk 2 above)
Comment (AWoloszyn) [reply gapil/runtime/cc/map.inc]: Done

File: gapil/runtime/cc/map.inc (hunk 1 above)
Comment (AWoloszyn) [reply gapil/runtime/cc/map.inc]: I think actually both end up being wrong for the same reason. (memory alignment :\) if alignof(float) != alignof(uint32_t).

I will do the "right thing" and go through a memcpy, which as far as I can tell is the only safe way to do this in general.

File: gapil/runtime/cc/map.inc (hunk 3 above)
Comment (AWoloszyn) [reply gapil/runtime/cc/map.inc]: Done

**Pull request merged**